### PR TITLE
Feature/22445

### PR DESCRIPTION
**Pull request by rochamarcelo**

### CODE REVIEW

File: src/Middleware/OneTimePasswordAuthenticatorMiddleware.php [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+     */
+    public function __invoke(ServerRequest $request, ResponseInterface $response, $next)
+    {
+        $service = $request->getAttribute('authentication');
+
+        if (!$service->getResult() || $service->getResult()->getStatus() !== AuthenticationService::NEED_TWO_FACTOR_VERIFY) {
+            return $next($request, $response);
+        }
+
+        $request->getSession()->write('CookieAuth', [
```
Comment (steinkel) [inline src/Middleware/OneTimePasswordAuthenticatorMiddleware.php]: extract CookieAuth into constant, as it's being used in 2 files at least

File: src/Authenticator/CookieAuthenticator.php [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    /**
+     * {@inheritDoc}
+     */
+    public function persistIdentity(ServerRequestInterface $request, ResponseInterface $response, $identity)
+    {
+        $field = $this->getConfig('rememberMeField');
+
+        $bodyData = $request->getParsedBody();
+        if (empty($bodyData)) {
+            $session = $request->getAttribute('session');
```
[the remark this answers is not in the document]
Comment (rochamarcelo) [reply src/Authenticator/CookieAuthenticator.php]: We could, I didn't used getSession cause it is not defined in Psr\Http\Message\ServerRequestInterface

File: src/Authenticator/FormAuthenticator.php [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+    public function authenticate(ServerRequestInterface $request, ResponseInterface $response)
+    {
+        $result = $this->getBaseAuthenticator()->authenticate($request, $response);
+        $checkKey = $this->getConfig('keyCheckEnabledRecaptcha');
+        if (!Configure::read($checkKey) || in_array($result->getStatus(), [Result::FAILURE_OTHER, Result::FAILURE_CREDENTIALS_MISSING])) {
+            return $this->lastResult = $result;
+        }
+
+        $data = $request->getParsedBody();
+        $captcha = $data['g-recaptcha-response'] ? $data['g-recaptcha-response'] : null;
```
[the remark this answers is not in the document]
Comment (rochamarcelo) [reply src/Authenticator/FormAuthenticator.php]: Added validateReCaptchaFromRequest

File: src/Middleware/OneTimePasswordAuthenticatorMiddleware.php (hunk 1 above)
Comment (rochamarcelo) [reply src/Middleware/OneTimePasswordAuthenticatorMiddleware.php]: Done

File: src/Authenticator/CookieAuthenticator.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (rochamarcelo) [reply src/Authenticator/CookieAuthenticator.php]: Updated

File: src/Authenticator/FormAuthenticator.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (rochamarcelo) [reply src/Authenticator/FormAuthenticator.php]: Added method validateReCaptchaFromRequest

File: src/Authenticator/TwoFactorAuthenticator.php [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+        if ($data === null) {
+            return new Result(null, Result::FAILURE_CREDENTIALS_MISSING, [
+                'Login credentials not found'
+            ]);
+        }
+
+        $request->getSession()->delete(self::USER_SESSION_KEY);
+
+        return new Result($data, Result::SUCCESS);
```
[the remark this answers is not in the document]
Comment (rochamarcelo) [reply src/Authenticator/TwoFactorAuthenticator.php]: Update

**Pull request merged**